tfhe/src/integer/server_key/radix_parallel/ilog2.rs

```diff
@@ -0,0 +1,734 @@
+    /// assert_eq!(n, msg.leading_zeros());
+    /// ```
+    pub fn leading_zeros<T>(&self, ct: &T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        let mut tmp;
+        let ct = if ct.block_carries_are_empty() {
+            ct
+        } else {
+            tmp = ct.clone();
+            self.full_propagate_parallelized(&mut tmp);
+            &tmp
+        };
+        self.unchecked_leading_zeros(ct)
+    }
+
+    /// Returns the number of leading ones in the binary representation of `ct`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2;
+    ///
+    /// // Generate the client key and the server key:
+    /// let num_blocks = 4;
+    /// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2, num_blocks);
+    ///
+    /// let msg = -4i8;
+    ///
+    /// let ct1 = cks.encrypt_signed(msg);
+    ///
+    /// let n = sks.leading_ones(&ct1);
+    ///
+    /// // Decrypt:
+    /// let n: u32 = cks.decrypt(&n);
+    /// assert_eq!(n, msg.leading_ones());
+    /// ```
+    pub fn leading_ones<T>(&self, ct: &T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        let mut tmp;
+        let ct = if ct.block_carries_are_empty() {
+            ct
+        } else {
+            tmp = ct.clone();
+            self.full_propagate_parallelized(&mut tmp);
+            &tmp
+        };
+        self.unchecked_leading_ones(ct)
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests_unsigned {
+    use super::*;
+    use crate::integer::keycache::KEY_CACHE;
+    use crate::integer::server_key::radix_parallel::tests_cases_unsigned::{
+        random_non_zero_value, FunctionExecutor, NB_CTXT, NB_TESTS_SMALLER,
+    };
+    use crate::integer::{IntegerKeyKind, RadixClientKey};
+    use crate::shortint::PBSParameters;
+    use rand::Rng;
+    use std::sync::Arc;
+
+    fn default_test_count_consecutive_bits<P, T>(
+        direction: Direction,
+        bit_value: BitValue,
+        param: P,
+        mut executor: T,
+    ) where
+        P: Into<PBSParameters>,
+        T: for<'a> FunctionExecutor<&'a RadixCiphertext, RadixCiphertext>,
+    {
+        let (cks, mut sks) = KEY_CACHE.get_from_params(param, IntegerKeyKind::Radix);
+        let cks = RadixClientKey::from((cks, NB_CTXT));
+
+        sks.set_deterministic_pbs_execution(true);
+        let sks = Arc::new(sks);
+
+        let mut rng = rand::thread_rng();
+
+        // message_modulus^vec_length
+        let modulus = cks.parameters().message_modulus().0.pow(NB_CTXT as u32) as u64;
+
+        executor.setup(&cks, sks.clone());
+
+        let num_bits = NB_CTXT as u32 * cks.parameters().message_modulus().0.ilog2();
+
+        let compute_expected_clear = |x: u64| match (direction, bit_value) {
+            (Direction::Trailing, BitValue::Zero) => {
+                if x == 0 {
+                    num_bits
```

Comment:
does trailing zeros fail here ?

Reply:
yes 0u64.trailing_zeros() = 64, not num_bits

Reply:
ah right indeed thanks